Create a swap space for installing NW

smpsgill opened the pull request:
Initial version TODO:
- Create idempotent states
- Check and add better conditions based on the requirements of SAP installation(NW, HANA and so on) and if each host needs to have the swap space

Commit:
Create a swap space for installing NW

arbulu89 commented on netweaver/setup/swap_space.sls:
```diff
@@ -0,0 +1,37 @@
+{%- from "netweaver/map.jinja" import netweaver with context -%}
+{% set host = grains['host'] %}
+
+{% for node in netweaver.nodes if host == node.host %}
```
I would say this for loop is not needed as this operation must be applied to all nodes (actually, this is what will happen with the current implementation, but it's more confusing)

arbulu89 commented on netweaver/setup/swap_space.sls:
```diff
@@ -0,0 +1,37 @@
+
+{% for node in netweaver.nodes if host == node.host %}
+
+{% set swap_file = '/var/lib/swap/swapfile' %}
```
Should we add this as a variable in the pillar file? And put this value as default.
Just wondering if this makes sense or not.
Otherwise this hardcoded value looks good

arbulu89 commented on netweaver/setup/swap_space.sls:
```diff
@@ -0,0 +1,37 @@
+
+{% set swap_file = '/var/lib/swap/swapfile' %}
+
+create_swap_dir:
```
We should use the `file.directory` state to make this idempotent instead of using `cmd.run`

arbulu89 commented on netweaver/setup/swap_space.sls:
```diff
@@ -0,0 +1,37 @@
+  cmd.run:
+  - name: mkdir -p /var/lib/swap
+
+create_swap_file:
```
We should not run the command if the file is already there. Add and `unless/onlyIf` option there checking if the file exist

smpsgill replied:
Yes, the creates will check this. It will Only run if the file or files specified by creates do not exist.

arbulu89 reviewed (changes requested):
In general I like the implementation.
Some comments added.
Maybe we should add a variable to the end of the state names just to avoid potential issues with duplicated names. As example:
https://github.com/salt-formulas/salt-formula-linux/blob/master/linux/storage/swap.sls

Some more suggestions:
- Maybe we should give the option to enable/disable swap creation via pillar. Something like `create_swap: True` being True by default.
- Give the option to select the swap file and size. I don't understand this mem_total * 2 approach. We should have a default value, but provide the option to select size too.

Commit:
Make improvemts to swap logic
- Added hostname to state names
- Make optional in pillar to create swap
- Set defaults to create_swap, file and size
- Remove looping

smpsgill commented:
Hi @arbulu89 I made improvements as per suggestion.
I am also thinking that this state of creating swap should go into our future SAP formula to setup the SAP infrastructure.

arbulu89 commented on netweaver/setup/swap_space.sls:
```diff
@@ -0,0 +1,38 @@
+
+{% if netweaver.create_swap|default(true) %}
+{% set swap_dir = netweaver.swap_config.directory|default('/var/lib/swap', true) %}
+{% set swap_file =  swap_dir~'/swapfile' %}
```
Double space after `=`

arbulu89 commented on pillar.example:
```diff
@@ -9,7 +9,11 @@ netweaver:
   swpm_folder: your_swpm_folder_relative_to_swpm_media
   sapexe_folder: your_sapexe_sar_folder_relative_to_swpm_media
-
+  create_swap: True
```
What do you think if we just skip the `create_swap: True` and put directly and replace the `swap_config` by that parameter.
Just to remove one dependant parameter. The final result would be:
```
create_swap:
      directory: /var/lib/swap
      size: 128
```
I think this would be more robust.

arbulu89 replied:
Please, if you add these changes document them in the pillar.example file. Say how the parameters have to be used

arbulu89 commented on pillar.example:
```diff
@@ -9,7 +9,11 @@ netweaver:
+  create_swap: True
+  swap_config:
+      directory: /var/lib/swap
+      size: 128
```
We should add a comment here. Which is the unit of the size?

arbulu89 reviewed (changes requested):
Some minor comments here.
Besides that the implementation is good and it's ready to merge

Commit:
Minor changes to swap pillar data

arbulu89 commented on netweaver/setup/swap_space.sls:
```diff
@@ -0,0 +1,38 @@
+{%- from "netweaver/map.jinja" import netweaver with context -%}
+{% set host = grains['host'] %}
+
+{% if netweaver.create_swap|default(true) %}
```
@Simranpal, you forgot to update the code after updating the pillar example here

smpsgill replied:
Hi @arbulu89 yes the existing logic did not work as such for all cases, I have updated it to consider more conditions

arbulu89 commented on pillar.example:
```diff
@@ -9,7 +9,10 @@ netweaver:
   swpm_folder: your_swpm_folder_relative_to_swpm_media
   sapexe_folder: your_sapexe_sar_folder_relative_to_swpm_media
-
+  swap_config:
```
Please add a comment saying how to use the `swap_config` option. Basically, if this entry exist, the swap partition will be created using the swap configuration parameters.

arbulu89 reviewed (changes requested):
@Simranpal, you forget to make some changes

arbulu89 commented on netweaver/setup/swap_space.sls:
```diff
@@ -0,0 +1,38 @@
+{% if netweaver.create_swap|default(true) %}
+{% set swap_dir = netweaver.swap_config.directory|default('/var/lib/swap', true) %}
+{% set swap_file = swap_dir~'/swapfile' %}
+{% set swap_size = netweaver.swap_config.size|default(128, true) %}
```
I would put 2048 by default, as this is the requited value by SAP

Commit:
Improve logic to consider all cases
& add documentation to pillar example

Commit:
Remove usage for swap_config
- Replaced by create_swap
- Added custom values to swap config

smpsgill commented:
Hi @arbulu89 I tested these changes with various conditions provided in pillar. 
Let me know what you think of the additional logic.

arbulu89 reviewed (changes requested):
@Simranpal,
I don't understand the last changes... It was already good in the last commit (with just a minimal change).

All this logic is super confusing in my opinion. We have added extra complexity with this new if/else statement.

Let me suggest other solution.

If we want to enable the swap creation by default (that might be a good idea), let's add this entry to the [defaults.yaml file](https://github.com/SUSE/sapnwbootstrap-formula/blob/master/netweaver/defaults.yaml). There we can add the default `swap_create` data after netweaver (actually, you can even remove the `version` entry, it's not used.

And then just check if the `create_swap` is defined and it's not False.

Something like:

```
# If create_swap entry exists and it's not False
{% if netweaver.create_swap is defined and netweaver.create_swap %}
{% set swap_dir = netweaver.create_swap.directory %}
{% set swap_size = netweaver.create_swap.size %}
# The rest of the state file
```

smpsgill commented:
> @Simranpal,
> I don't understand the last changes... It was already good in the last commit (with just a minimal change).
> 
> All this logic is super confusing in my opinion. We have added extra complexity with this new if/else statement.
> 
> Let me suggest other solution.
> 
> If we want to enable the swap creation by default (that might be a good idea), let's add this entry to the [defaults.yaml file](https://github.com/SUSE/sapnwbootstrap-formula/blob/master/netweaver/defaults.yaml). There we can add the default `swap_create` data after netweaver (actually, you can even remove the `version` entry, it's not used.
> 
> And then just check if the `create_swap` is defined and it's not False.
> 
> Something like:
> 
> ```
> # If create_swap entry exists and it's not False
> {% if netweaver.create_swap is defined and netweaver.create_swap %}
> {% set swap_dir = netweaver.create_swap.directory %}
> {% set swap_size = netweaver.create_swap.size %}
> # The rest of the state file
> ```

Hi @arbulu89 you are correct. I think I overcomplicated the logic a bit. I will revert to previous logic and you suggestion. Thank you!

Commit:
Revert to simple logic
- Added swap creation to defaults.yaml file
- Remove swap entries from pillar.example

arbulu89 reviewed (approved):
Thank you for the changes.
I think is ready to be merged now
